Database schema init & migrations

Light-it-luc opened the pull request:
# Light-it Pull Request Template ⚡⚡⚡

## Description ⭐

I've created the first iteration of the main database models (`Airline`, `City` and `Flight`),
along with the corresponding migrations (including `airline_city` pivot table) and factory classes.

### Type of change 🏃

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [x] This change requires a documentation update

### Requires ❗

- [x] This change requires a database update or migration
- [x] This change requires a documentation update


## Emoji Guide

**For reviewers: Emojis can be added to comments to call out blocking versus non-blocking feedback.**

E.g., Praise, minor suggestions, or clarifying questions that don’t block merging the PR.

> 🟢 Nice refactor!
> 🟡 Why was the default value removed?

E.g., Blocking feedback must be addressed before merging.

> 🔴 This change will break something important

|              |                |                                     |
| ------------ | -------------- | ----------------------------------- |
| Blocking     | 🔴 ❌ 🚨       | RED                                 |
| Non-blocking | 🟡 💡 🤔 💭    | Yellow, thinking, etc               |
| Praise       | 🟢 💚 😍 👍 🙌 | Green, hearts, positive emojis, etc |

## Links

- [Git Flow](https://lightit.slite.com/app/docs/SC8usN2Ju)
- [Handbook of good practices for reviewers in Code Reviews](https://lightit.slite.com/app/docs/ddNGohWthVB3fO)

LCobham committed:
feat: Create first iteration of models Airline, City and Flight

LCobham committed:
feat: Create migration files for tables Airline, City, Flight & airline_city

LCobham committed:
feat: Create first iteration of factories for main models and DB seeder

LCobham committed:
refactor: Simplify Flight factory

guillemartinicore commented on database/factories/AirlineFactory.php:
```diff
@@ -0,0 +1,24 @@
+    public function definition(): array
+    {
+        return [
+            'name' => $this->faker->company() . ' Airline',
```
Don't you think this should be '->unique()' as well?

guillemartinicore commented on database/migrations/2023_11_21_163804_create_airlines_table.php:
```diff
@@ -0,0 +1,29 @@
+    {
+        Schema::create('airlines', function (Blueprint $table) {
+            $table->id();
+            $table->string('name', 255);
```
same here, name should be unique as it's the only way of differentiate one airline from another. It's a requirement for you to check that an airline is not repeated.

guillemartinicore commented on database/migrations/2023_11_21_163810_create_flights_table.php:
```diff
@@ -0,0 +1,34 @@
+            $table->foreignId('airline_id')->constrained()->onDelete('cascade');
+            $table->foreignId('origin_city_id')->references('id')->on('cities')->onDelete('cascade');
+            $table->foreignId('dest_city_id')->references('id')->on('cities')->onDelete('cascade');
+            $table->dateTimeTz('departure');
```
you should name this as 'departure_at'

guillemartinicore commented on database/migrations/2023_11_21_163810_create_flights_table.php:
```diff
@@ -0,0 +1,34 @@
+            $table->foreignId('origin_city_id')->references('id')->on('cities')->onDelete('cascade');
+            $table->foreignId('dest_city_id')->references('id')->on('cities')->onDelete('cascade');
+            $table->dateTimeTz('departure');
+            $table->dateTimeTz('arrival');
```
same here

guillemartinicore commented on database/migrations/2023_11_21_163810_create_flights_table.php:
```diff
@@ -0,0 +1,34 @@
+            $table->foreignId('dest_city_id')->references('id')->on('cities')->onDelete('cascade');
+            $table->dateTimeTz('departure');
+            $table->dateTimeTz('arrival');
+            $table->boolean('active')->default(true);
```
what do you want to do with this?

guillemartinicore commented on database/factories/FlightFactory.php:
```diff
@@ -0,0 +1,34 @@
+     */
+    public function definition(): array
+    {
+        $departure = now()->addDays(rand(0, 7));
```
Carbon instances could be useful for you here. Take a closer look https://carbon.nesbot.com/docs/ It will make you code cleaner and will keep you from using copy

LCobham committed:
refactor: Improvement in airline migration and flight factory

Add unique constraint on airline name & refactor of departure/arrival date in flight factory

LCobham committed:
refactor: name changes and small improvements in Flight migration & factory

Add previx `_at` to `departure` and `arrival`. Delete `active` variable

ezef commented on app/Models/Airline.php:
```diff
@@ -0,0 +1,23 @@
+{
+    use HasFactory;
+
+    protected $guarded = [];
```
always is good practise to guard the id

`protected $guarded = ['id'];`

ezef commented on app/Models/City.php:
```diff
@@ -0,0 +1,30 @@
+
+class City extends Model
+{
+    use HasFactory;
```
In Light-it we are not really using the `HasFactory` trait on models, we just use `CityFactory::new()->create()`. This is because we do not want the Domain model to know about his factory ( S from Solid, single responsability principle)

ezef commented on app/Models/Flight.php:
```diff
@@ -0,0 +1,43 @@
+        }
+
+        return $flight_n;
+    }
```
This is quite difficult to understand, you have in Laravel Str helpers, maybe you can take a look at them and refactor it https://laravel.com/docs/10.x/strings

ezef commented on database/factories/AirlineFactory.php:
```diff
@@ -0,0 +1,24 @@
+            'name' => $this->faker->unique()->company() . ' Airline',
+            'description' => $this->faker->paragraph()
+        ];
+    }
```
If you remove the HasFactory trait on `Airline` model, put `protected $model = Airline::class;` here. Also do it on all Factories

ezef commented on database/migrations/2023_11_21_163810_create_flights_table.php:
```diff
@@ -0,0 +1,33 @@
+            $table->foreignId('airline_id')->constrained()->cascadeOnDelete();
+            $table->foreignId('origin_city_id')->constrained('cities')->cascadeOnDelete();
+            $table->foreignId('dest_city_id')->constrained('cities')->cascadeOnDelete();
+            $table->dateTimeTz('departure_at');
+            $table->dateTimeTz('arrival_at');
```
let's just use `->dateTime`. We should always use dateTime and consider it UTC, unless the use case requires to have TZ on DB

ezef commented on database/seeders/DatabaseSeeder.php:
```diff
@@ -12,11 +12,6 @@ class DatabaseSeeder extends Seeder
-        //     'name' => 'Test User',
-        //     'email' => 'test@example.com',
-        // ]);
+        Flight::factory(10)->create();
```
After the changes about Factories, you can use here `FlightFactory::new()->count(10)->create();`

LCobham committed:
refactor: Remove dependency between models & factories. Guard ids in models

LCobham committed:
refactor: Simplify generator of random flight numbers

LCobham committed:
refactor: Change flights migration to remove TZ from datetime vars

LCobham committed:
style: Modify flight number generator style

LCobham committed:
refactor: Simplify flight number generator
